Upgrade and generalise basis state preparation (#6021)
PennyLaneAI/pennylane: doc/development/deprecations.rst

  - Deprecated in v0.37
  - Will be removed in v0.39

* The ``BasisStatePreparation`` template is deprecated.
  Instead, ``BasisState`` can be called on the constructed operator.

  - Deprecated in v0.38

>   - Will be removed in v0.39

Author: Double check

Reviewer: Works for me! Any concerns @isaacdevlugt?

Reviewer 2: Thanks @KetpuntoG! I think it's fine to deprecate this for v0.38, remove for v0.39. I created a story in the deps / removals for v0.38 that relates to the story Guillermo is working on :)

  - Deprecated in v0.37
  - Will be removed in v0.39

> * The ``BasisStatePreparation`` template is deprecated.
>   Instead, ``BasisState`` can be called on the constructed operator.

Reviewer 2: ```suggestion
* The ``BasisStatePreparation`` template is deprecated.
  Instead, use ``BasisState``.
```
